cluster-autoscaler/cloudprovider/aws/auto_scaling_groups.go

```diff
@@ -141,17 +141,21 @@ func (m *asgCache) register(asg *asg) *asg {
 
 // Unregister ASG. Returns the unregistered ASG.
 func (m *asgCache) unregister(a *asg) *asg {
-	updated := make([]*asg, 0, len(m.registeredAsgs))
 	var changed *asg
+	var i int
 	for _, existing := range m.registeredAsgs {
 		if existing.AwsRef == a.AwsRef {
 			klog.V(1).Infof("Unregistered ASG %s", a.AwsRef.Name)
 			changed = a
 			continue
 		}
-		updated = append(updated, existing)
+		m.registeredAsgs[i] = existing
```

Comment:
I might be mistaken but I assumed the previous version was copying/replacing (but not updating in place) because the cluster-autoscaler don't lock-protect that cache's content (slice items).

regenerate(), register() and unregister() operates under m.mutex.Lock and regenerates a new m.registeredAsgs (the containing slice), while other threads might have obtained an older version via asgCache.Get() and still access a stale/previous m.registeredAsgs versions (which should be gc'ed later).

If that's the case then updating in place wouldn't be safe; what do you think?

Reply:
Thanks for reviewing.

Yes, you are right, although `asgCache` is guarded by `mutex.Lock`, the returned `registeredAsgs` is out of the lock's control. So I can't just reuse the slice here.